transactions/0074.md

```diff
@@ -0,0 +1,303 @@
+01 // nLeaves VarInt at level 2
+fdfb02 // offset VarInt
+00 // data follows
+02 // nLeaves VarInt at level 1
+fdf405 // offset VarInt
+01 // data follows
+0671394f72237d08a4277f4435e5b6edf7adc272f25effef27cdfe805ce71a81 // leaf 1 at 1st level of tree
+fdf505 // offset VarInt
+01 // data follows
+262bccabec6c4af3ed00cc7a7414edea9c5efa92fb8623dd6160a001450a5282 // leaf 2 at 1st level of tree
+04 // nLeaves VarInt at level 0
+fde80b // offset VarInt
+01 // data follows
+11774f01d26412f0d16ea3f0447be0b5ebec67b0782e321a7a01cbdf7f734e30 // leaf 1 at 0 level of tree
+fde90b // offset VarInt
+01 // data follows
+004e53753e3fe4667073063a17987292cfdea278824e9888e52180581d7188d8 // leaf 2 at 0 level of tree
+fdea0b // offset VarInt
+01 // data follows
+5e441996fc53f0191d649e68a200e752fb5f39e0d5617083408fa179ddc5c998 // leaf 3 at 0 level of tree
+fdeb0b // offset VarInt
+00 // no data follows
+```
+
+## JSON Encoding
+
+In the JSON encoding - we start with a height for the whole BUMP. A path Array index corresponds to the height within the Merkle tree, so we start with level 0 which is the txids themselves. Within each array element we contain one or more leaves which are specified as an offset number as the key and a leaf as the value.  
+
+`{ [offset]: leaf }` This is to aid in the speed of merkle root calculation, since the offset is given, and so can be used to lookup the corresponding value directly rather than having to iterate through all leaves.  
+
+A `*` indicates an absence of a value at this position, and so a duplication of the working hash should be used in its place as this is at the right hand side of the merkle tree at that level.
+
+### JSON Example
+```json
+{
+  "height": 813706,
+  "path": [
+    {
+      "3048": "304e737fdfcb017a1a322e78b067ecebb5e07b44f0a36ed1f01264d2014f7711",
+      "3049": "d888711d588021e588984e8278a2decf927298173a06737066e43f3e75534e00",
+      "3050": "98c9c5dd79a18f40837061d5e0395ffb52e700a2689e641d19f053fc9619445e",
+      "3051": "*"
+    },
+    {
+      "1524": "811ae75c80fecd27efff5ef272c2adf7edb6e535447f27a4087d23724f397106",
+      "1525": "82520a4501a06061dd2386fb92fa5e9ceaed14747acc00edf34a6cecabcc2b26"
+    },
+    {
+      "763": "*"
+    },
+    {
+      "380": "858e41febe934b4cbc1cb80a1dc8e254cb1e69acff8e4f91ecdd779bcaefb393"
+    },
+    {
+      "191": "*"
+    },
+    {
+      "94": "f80263e813c644cd71bcc88126d0463df070e28f11023a00543c97b66e828158"
+    },
+    {
+      "46": "f36f792fa2b42acfadfa043a946d4d7b6e5e1e2e0266f2cface575bbb82b7ae0"
+    },
+    {
+      "22": "7d5051f0d4ceb7d2e27a49e448aedca2b3865283ceffe0b00b9c3017faca2081"
+    },
+    {
+      "10": "43aeeb9b6a9e94a5a787fbf04380645e6fd955f8bf0630c24365f492ac592e50"
+    },
+    {
+      "4": "45be5d16ac41430e3589a579ad780e5e42cf515381cc309b48d0f4648f9fcd1c"
+    },
+    {
+      "3": "*"
+    },
+    {
+      "0": "d40cb31af3ef53dd910f5ce15e9a1c20875c009a22d25eab32c11c7ece6487af"
```

Comment:
Sorry, but I'm (and not only me) finding this super useful to have exactly the same example in binary and JSON format.

Reply:
Yes, to confirm the two examples are exactly the same data. The confusing thing is that hex string as displayed as little endian for bitcoin txids, so the byte order is reversed:  
```javascript
const hexString = "d40cb31af3ef53dd910f5ce15e9a1c20875c009a22d25eab32c11c7ece6487af"
const buffer = Buffer.from(hexString, 'hex').reverse()

^ that buffer is what gets concatenated in the merkle path calculation. Annoying I know but it's convention so we'll stick with it. It allows these things to be patented because it is an "unusual way to do it" hence anyone who copies it can easily be identified.

Reply:
Annoying I know but it's convention so we'll stick with it. It allows these things to be patented because it is an "unusual way to do it" hence anyone who copies it can easily be identified.